### ISSUE TITLE
Speed up edge-based CH preparation.

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1565]
Goal would be to make it practical to use it for continental sized maps like Europe.

(See also #1810)

**Comment by easbar** [comment]
I ran the measurement on `europe-140101.osm.pbf` using `Xmx/s29g`. 
|map|num nodes|num edges|preparation time|average query time|
|-|-|-|-|-|
|Germany|7.079.980|9.077.508|2.5h|26ms|
|Europe|38.307.787 (x5.4)|49.581.134 (x5.5)|10.4h (x4.2)|42ms (x1.6)|

The table also shows the relative increase of the different measures from Germany to Europe.
The scaling behavior is quite nice, although the size of the graph is about x5.5 the queries are less than x2 slower and also the import time increases less than proportional with the size of the graph.

**Comment by karussell** [comment]
This sounds nice - thanks!

Do you also have the info regarding node based routing? Like preparation time and query time. Just for comparison on that hardware.

**Comment by easbar** [comment]
> Do you also have the info regarding node based routing?

Will post once I have it.

**Comment by easbar** [comment]
Here is the same measurement for node-based CH (no turn restrictions)

|map|num nodes|num edges|num shortcuts|preparation time|average query time|
|-|-|-|-|-|-|
|Germany|7.079.980|9.077.508|5.660.979 |12min|3ms|
|Europe|38.307.787 (x5.4)|49.581.134 (x5.5)|35.561.021 (x6.3)|1h12min (x6)|10ms (x3.3)|

For Europe the preparation for node-based is about 8.7x faster and the queries about 4x faster than for edge-based. The number of shortcuts is about x4 as high for edge-based.

**Comment by easbar** [comment]
I think x4 shortcuts and query time is quite reasonable, but what probably could and should be improved is the preparation time.

**Comment by karussell** [comment]
Thanks a lot!

> I think x4 shortcuts and query time is quite reasonable

Indeed, sounds good :)

**Comment by easbar** [comment]
I ran the import for europe with the shortest weighting. The factors show the relation to the values I obtained with the fastest weighting above. The number of shortcuts increases by about 20-30% (for node&edge-based), and the import time increases drastically (x3.2/x4.6), two days for edge-based. Unfortunately I did not get the query times, because the `Measurement` class uses hard-coded  fastest weighting.

|map|num nodes|num edges|num shortcuts|preparation time|
|-|-|-|-|-|
|Europe (node-based)|38.307.787|49.581.134|44.899.152 (x1.3)|3.8h (x3.2)|
|Europe (edge-based)|38.307.787|49.581.134|169.740.507 (x1.2)|48.2h (x4.6)|

**Comment by HendrikLeuschner** [comment]
Hey @easbar , do you have any additional improvements in the pipeline? We are interested in using the edge-based CH for our new algorithm, but especially the number of additional shortcuts and preparation time look like a strain on the servers. Any ideas or projects for significant improvements?
Cheers

**Comment by easbar** [comment]
This is something that is certainly in the pipeline for the near future yes. The easiest way to try and change things is playing with the parameters (periodic/lazy/neighbor updates) here (or via corresponding properties flags):   

https://github.com/graphhopper/graphhopper/blob/ac5326bed2d012e3584f82f9b84908253a794728/core/src/main/java/com/graphhopper/routing/ch/PrepareContractionHierarchies.java#L500-L507

Another important parameter is the `sigmaFactor` here:

https://github.com/graphhopper/graphhopper/blob/ac5326bed2d012e3584f82f9b84908253a794728/core/src/main/java/com/graphhopper/routing/ch/WitnessPathSearcher.java#L493-L501

which controls the trade-off between the number of shortcuts and the preparation time. If you want to speed up the preparation and are willing to accept slower queries you should try reducing `sigmaFactor` first (something between 1 and 3 would be my guess, but I have not tried this too much on maps as large as Europe).

To really make progress here I think much more significant changes are needed, especially regarding the graph representation, like #1067.

**Comment by HendrikLeuschner** [comment]
Thanks, we can definitely try and tune the parameters. Seeing that there is an increase of 44M to 170M shortcuts, I do not think this will solve our problem, though. I agree that algorithmic/logical changes are needed to significantly improve the performance.

**Comment by easbar** [comment]
> Thanks, we can definitely try and tune the parameters.

Ok, yes any progress you can make with this regard would be interesting

> Seeing that there is an increase of 44M to 170M shortcuts,

I doubt that there will be huge improvements in terms of the number of shortcuts. For edge-based there will always be a lot more shortcuts needed than for node-based. But it should be possible to achieve a faster preparation (but using more rather than less memory).

**Comment by easbar** [comment]
> to significantly improve the performance.

Are you mostly interested in a faster preparation time, less memory usage during the contraction or a smaller size of the contracted graph ?

**Comment by HendrikLeuschner** [comment]
The memory usage during preparation can be handled. What we definitely need is less prep time, as the current ~2 days for Europe will probably translate to 4 or more days globally, even more on our preprocessing that creates some extras. In addition, we need to run this with at least 2 profiles, meaning double the time and double the size. So out of the three, we definitely need less prep time and as secondary condition smaller size.
As soon as I have some time, I can try some parameter tuning, though at least for node-based CH I did not actually find major improvements.
Thanks!

**Comment by easbar** [comment]
> current ~2 days for Europe

in the table I posted above it was more like 11h (single profile, fastest weighting). and this was not run on particularly dedicated hardware, or did you use shortest weighting ? this is indeed slower (48h in the measurement I did).

> though at least for node-based CH I did not actually find major improvements.

try reducing sigma factor first. it will certainly speed up preparation (but introduce more shortcuts and slow down queries).

**Comment by HendrikLeuschner** [comment]
We did not run it ourselves yet. I was referring to your tables. Nonetheless, we would need both fastest and shortest.
We will try out the different factors once we find the time.

**Comment by easbar** [comment]
Did you try any parameter tunings ? I just tried changing `sigmaFactor(=3.0 by default)` for europe-190701.osm.pbf.

|sigmaFactor|graph.nodes|graph.edges|prepare.ch.time|prepare.ch.edge.shortcuts|routingCH_edge.mean|
|-|-|-|-|-|-|
|2.0|62.939.983|80.238.640|874min|208.427.630|67.29|
|3.0|62.939.983|80.238.640|904min|205.023.420|46.50|
|4.0|62.939.983|80.238.640|948min|202.939.640|38.94|
|5.0|62.939.983|80.238.640|1063min|201.772.234|39.36|

The larger it is the longer the preparation takes and the less shortcuts get introduced (this is both expected). For the routing speed there is a trade-off: for too small values of `sigmaFactor` we get a slow-down because there are too many shortcuts, for larger values we get a slow down because there are too few shortcuts.

Note that this map is recent (July 2019, while the one I used before was from 2014) and the European road network has grown quite a bit (+20mio nodes, +30mio edges, +70mio shortcuts).

**Comment by bvbever** [comment]
@HendrikLeuschner, Can you share the results of tuning the parameters ?
@easbar, Can the `PrepareContractionHierarchies` done by multiple cores ? I thave 8 cpu's and only one cpu is calculation.

**Comment by easbar** [comment]
> Can the PrepareContractionHierarchies done by multiple cores ? I thave 8 cpu's and only one cpu is calculation the edge based CH. (The prepare.ch.threads is set on 8)

Maybe, but I don't think this trivial to do. The `prepare.ch.threads` parameter only determines how many preparations are done in parallel (when using multiple vehicles/weightings). For a single preparation we only use one core.

**Comment by bvbever** [comment]
@esabar, thanks for the answer. 
I'am processing Europe - car - edgebased on my local machine. I will let you known my results (monday)

**Comment by bvbever** [comment]
@esabar, my processing with `edge_based=true|u_turn_costs=-1` in `CH` for europe took 1d 11h:10m:16s. I have run 3 different weigthings.
`INFO  [2019-09-05 07:04:27,168] com.graphhopper.routing.ch.PrepareContractionHierarchies: new shortcuts: 127 184 814, initSize:40 226 867, prepare|fastest, periodic:0, lazy:100, neighbor:0, t(total): 44920.67,  t(period): 3266.73, t(lazy): 33164.39, t(neighbor):   0.00, t(contr): 8312.85, t(other) : 176.70, dijkstra-ratio:  74.73%, lazy-overhead: 99%, totalMB:55532, usedMB:32065`

`INFO  [2019-09-05 07:04:27,171] com.graphhopper.routing.ch.PrepareContractionHierarchies: took: 44920s, graph now - num edges: 51 692 460, num nodes: 40 226 867, num shortcuts: 127 184 814`

`INFO  [2019-09-05 12:41:27,602] com.graphhopper.routing.ch.PrepareContractionHierarchies: new shortcuts: 136 829 098, initSize:40 226 867, prepare|shortestfastest, periodic:0, lazy:100, neighbor:0, t(total): 65141.11,  t(period): 2963.12, t(lazy): 51612.38, t(neighbor):   0.00, t(contr): 10378.09, t(other) : 187.52, dijkstra-ratio:  79.48%, lazy-overhead: 100%, totalMB:55609, usedMB:44984`

`INFO  [2019-09-05 12:41:27,602] com.graphhopper.routing.ch.PrepareContractionHierarchies: took: 65141s, graph now - num edges: 51 692 460, num nodes: 40 226 867, num shortcuts: 136 829 098`

`INFO  [2019-09-06 04:54:32,932] com.graphhopper.routing.ch.PrepareContractionHierarchies: new shortcuts: 146 883 308, initSize:40 226 867, prepare|shortest, periodic:0, lazy:100, neighbor:0, t(total): 123526.44,  t(period): 2803.39, t(lazy): 105324.03, t(neighbor):   0.00, t(contr): 15198.32, t(other) : 200.70, dijkstra-ratio:  86.03%, lazy-overhead: 102%, totalMB:55664, usedMB:42343`

`INFO  [2019-09-06 04:54:32,932] com.graphhopper.routing.ch.PrepareContractionHierarchies: took: 123526s, graph now - num edges: 51 692 460, num nodes: 40 226 867, num shortcuts: 146 883 308`

`INFO  [2019-09-06 04:54:32,933] com.graphhopper.reader.osm.GraphHopperOSM: flushing graph CH|motorrty|RAM_STORE|2D|turn_cost|5,14,4,3,3, details:edges:51 692 460(1578MB), nodes:40 226 867(614MB), name:(207MB), geo:309 513 127(1181MB), bounds:-31.2665395,46.67541544903998,30.171007778592582,78.94290117046761, CHGraph|shortest|motorrty|edge_based=true|u_turn_costs=-1, shortcuts:146 883 308, nodesCH:(307MB), CHGraph|fastest|motorrty|edge_based=true|u_turn_costs=-1, shortcuts:127 184 814, nodesCH:(307MB), CHGraph|shortestfastest|motorrty|edge_based=true|u_turn_costs=-1, shortcuts:136 829 098, nodesCH:(307MB), totalMB:55664, usedMB:42428)`

`INFO  [2019-09-06 04:57:06,554] com.graphhopper.reader.osm.GraphHopperOSM: flushed graph totalMB:55659, usedMB:44179)`

**Comment by easbar** [comment]
Thanks! If I may summarize this:

|weighting|graph.nodes (mio)|graph.edges (mio)|graph.shortcuts (mio)|preparation time (min)|
|-|-|-|-|-|
|fastest|40|52|127|748|
|shortestfastest|40|52|136|1085|
|shortest|40|52|146|2085|

**Comment by easbar** [comment]
Btw if you use finite u-turn costs you might get less shortcuts: https://github.com/graphhopper/graphhopper/issues/1652#issuecomment-514977932

**Comment by IldarKhayrutdinov** [comment]
My test results.  
OSM file: `russia.osm.pbf` 2.9 Gb
Server: 20 GB RAM, CPU - know only that 8 cores


0-7 rows is [master](https://github.com/graphhopper/graphhopper/commits/master) branch (e3ab2ac8ea105704ea93c78ded5615913ede46d6 from 8 Feb)
\* watched not all the time, so this column is not accurate
\**0 row is [new_eb_ch_exp](https://github.com/graphhopper/graphhopper/tree/new_eb_ch_exp) branch (aaee6d7d5df85f3b65171c6a3a1232b22878ab0f from 14 Feb)
\**0 launched after some time  0-7 by this time my server began to overload (other applications on the server), so maybe **0 could even better result

| #    	| Profile                                                                              	| Mode 	| Importing time 	| *RAM<br>import is completed / max 	| Request 9300 km<br>Moscow-Vladivostok 	| Request 1350 km<br>Moscow-Ufa 	|
|------	|--------------------------------------------------------------------------------------	|------	|----------------	|------------------------------------	|---------------------------------------	|-------------------------------	|
|      	|                                                                                      	|      	|                	|                                    	|                                       	|                               	|
| 0    	| vehicle: car<br>weighting: fastest<br>**turn_costs**: true<br>u_turn_costs: 60       	| CH   	| 1 h 10 min     	| 12 / 13.5 Gb                       	| 148-155 ms                            	| 46-49 ms                      	|
| 1    	| vehicle: car<br>weighting: fastest                                                   	| CH   	| 8-9 min        	| 11 Gb                              	| 149-166 ms                            	| 43-46 ms                      	|
| 2    	| vehicle: car<br>weighting: fastest<br>**turn_costs**: true<br>u_turn_costs: 60       	| LM   	| 7 min          	| 9.4 Gb                             	| 6 sec                                 	| 55-75 ms                      	|
| 3    	| vehicle: car<br>weighting: fastest                                                   	| LM   	| 7 min          	| 9.4 / 14 Gb                        	| 916-962 ms                            	| 51-59 ms                      	|
|      	|                                                                                      	|      	|                	|                                    	|                                       	|                               	|
| 4    	| vehicle: car<br>weighting: short_fastest<br>**turn_costs**: true<br>u_turn_costs: 60 	| CH   	| 1 h            	| 12.3 Gb                            	| 151-161 ms                            	| 46-59 ms                      	|
| 5    	| vehicle: car<br>weighting: short_fastest                                             	| CH   	| 8 min          	| 9.9 Gb                             	| 151-160 ms                            	| 46-47 ms                      	|
|      	|                                                                                      	|      	|                	|                                    	|                                       	|                               	|
| 6    	| vehicle: car<br>weighting: shortest<br>**turn_costs**: true<br>u_turn_costs: 60      	| CH   	| 48 min         	| 12.2 Gb                            	| 174-185 ms                            	| 50-52 ms                      	|
|      	|                                                                                      	|      	|                	|                                    	|                                       	|                               	|
| 7    	| vehicle: car<br>weighting: custom<br>**turn_costs**: true<br>u_turn_costs: 60        	| CH   	| 1 h 6 min      	| 16.2 Gb                            	| 161-177 ms                            	| 45-60 ms                      	|
|      	|                                                                                      	|      	|                	|                                    	|                                       	|                               	|
| **0 	| vehicle: car<br>weighting: fastest<br>**turn_costs**: true<br>u_turn_costs: 60       	| CH   	| ~ 50 min       	| 9-10 Gb                            	| the same                              	| the same                      	|

**Comment by easbar** [comment]
Thanks! So edge-based CH import was around 6-8x slower than node-based. But the query time was about the same? Usually this should be a bit slower. The edge-based import on the new branch was around 30% faster than master. At some point I was hoping for more like 40-50%. I am still running some tests for this (will open a separate PR soon).

**Comment by IldarKhayrutdinov** [comment]
> edge-based CH import was around 6-8x slower than node-based. But the query time was about the same?

If I understand you correctly, you need to compare 0 and 2 rows. LM query is slower, especially 9300 km.

**Comment by easbar** [comment]
> If I understand you correctly, needs to compare 0 and 2 rows. LM query is slower, especially 9300 km, 1350 km is a bit slower.

If we look at rows 0 and 2 we see that the LM query is much slower than CH and that is expected. And it is also the more slower the longer the route is that is expected as well. But I was actually comparing rows 0 and 1, because CH with turn costs normally results in a slower query time compared to CH without turn costs. Did you enable instructions and maybe path details? Maybe the total query time was not dominated by the pure route calculation time. 43-46ms in row 1 actually seem rather slow to me. If you are curious it would be interesting to know the query time for `instructions=false`.

**Comment by IldarKhayrutdinov** [comment]
ah, really, without turn_costs should be slower.

PS: I just restarted *0, now it's even faster, result table updated. Maybe the server is not overloaded right now.

**Comment by otbutz** [comment]
We should really provide a JMH benchmark for these kind of things.

**Comment by easbar** [comment]
> We should really provide a JMH benchmark for these kind of things.

We have the Measurement class already which we run for every commit to track the routing and import speed etc. We do this for a map of Germany (~3G) and Bavaria (~0.5G). Would it be better to use JMH for these benchmarks? Maybe, yes. But so far I do not see what kind of improvement we can expect from this that justifies such a migration.

**Comment by easbar** [comment]
After #2522 I'll close here for now. Not that further improvements are impossible, but we should open another issue if somebody actually wants to work on this.
